### ISSUE TITLE
rviz/Gazebo Not Rendering 3d Space on Intel Laptop

### ISSUE THREAD

**Issue by kneave** [microsoft/wslg#455]
# Environment

```none
Windows build number: Microsoft Windows NT 10.0.22000.0
Your Distribution version: Ubuntu 20.04
Your WSLg version: 1.0.26
```

# Steps to reproduce

1. Install ROS Noetic  
2. run an instance of roscore in one terminal
3. run `rosrun rviz rviz`  or `rosrun gazebo_sim gazebo`
4. UI is displayed for the two applications but not the 3d rendered space.

**WSL logs**: 

[weston.log](https://github.com/microsoft/wslg/files/7156438/weston.log)
[pulseaudio.log](https://github.com/microsoft/wslg/files/7156439/pulseaudio.log)

versions.txt:
WSLg ( x86_64 ): 1.0.19+Branch.main.Sha.2a47bf20c239b523d7413f315cbc9ee60dfde831
Mariner: VERSION="1.0.20210224"
FreeRDP: 5f083fa0b97d433d6204985f6047886e29c1c61e
weston: 16de531f00aa3dfd17e0de74c8f49e9fd7cec617
pulseaudio: 2f0f0b8c3872780f15e275fc12899f4564f01bd5
mesa:

#  Expected behavior

The applications should render the robot/simulation or at least the 3d grid.

# Actual behavior

Gazebo:
![image](https://user-images.githubusercontent.com/733411/133133599-59daa246-f5df-4474-92d4-63140dc015ac.png)

rviz:
![image](https://user-images.githubusercontent.com/733411/133133747-d41b819a-9546-4f06-b42c-0ce1e64ebcf3.png)

This works fine on my intel/nvidia based machine but on my Intel laptop it consistently shows this behaviour where previously it worked fine.

CPU: Intel Core i7-8565U
GPU: Intel UHD Graphics 620

I've tried with the latest Intel drivers which claim to support WSLg and the newer beta drivers with the same behaviour.

**Comment by anup-das** [comment]
I have the same issue on Intel UHD graphics 630 on my core i7 9700k system. Blank output in glxgears. Works fine on my another system having discrete nvidia 1650 super card

**Comment by anilzeybek** [comment]
I also have the same problem

**Comment by anup-das** [comment]
For me arch seems to work on same setup. https://github.com/yuk7/ArchWSL

**Comment by anilzeybek** [comment]
it is probably using software rendering in arch wsl, so it is not distro specific: https://github.com/microsoft/wslg/issues/482#issuecomment-957014827

**Comment by westonNexben** [comment]
I have the same problem, but new tidbit of information: Software rendering, after installing the generic Intel driver but before I installed the updated Mesa, didn't display any GUI windows at all.

Dell XPS 7590 (also has a dedicated Nvidia GPU, but apparently we want integrated for WSLg)
Windows 11
Intel Driver: 30.0.101.1069
Ubuntu 20.04

`cat /mnt/wslg/versions.txt`

```
WSLg ( x86_64 ): 1.0.26+Branch.main.Sha.26ce2c09b86442f3c7f4f6462f770ed2afa76a25
Mariner: VERSION="1.0.20210224"
FreeRDP: b05321cd4e6a862aef76163a69db4e1910245736
weston: 46756d0e77e5c01b5995fbbee6f3ab0db9b30612
pulseaudio: 2f0f0b8c3872780f15e275fc12899f4564f01bd5
mesa:
```

`glxgears -info`

```
GL_RENDERER   = D3D12 (Intel(R) UHD Graphics 630)
GL_VERSION    = 3.1 Mesa 21.3.1 - kisak-mesa PPA
GL_VENDOR     = Microsoft Corporation
```

**Comment by westonNexben** [comment]
I think the title of this bug should be made more generic to clarify acceleration is not working on Intel UHD.  Seems like a fairly important thing to fix.

**Comment by valinet** [comment]
@westonNexben Well, as mentioned in #334, I also happen to be facing the same problem. Furthermore, I also happen to have it on the same platform as you, similarly a Dell XPS 15 7590, with the UHD 630 and GTX 1650 GPU as well. Hopefully a solution is found.

**Comment by KiloNovemberDelta** [comment]
Hello everyone,
I confirm, I have exactly the same issue. Gazebo is working well on WSL2 on my other computer with Nvidia, but showing a black screen with my main computer using Intel(R) Iris(R) Xe Graphics. 
Still trying to find a solution, good luck in your research !

**Comment by KiloNovemberDelta** [comment]
Regarding Gazebo, I have set "export LIBGL_ALWAYS_SOFTWARE=1" and then run gazebo, and I get now the GUI working well. 
I hope it may help people who had the same issue.

**Comment by anilzeybek** [comment]
> Regarding Gazebo, I have set "export LIBGL_ALWAYS_SOFTWARE=1" and then run gazebo, and I get now the GUI working well. I hope it may help people who had the same issue.

yes but in this case you are not utilizing gpu, and since gazebo requires gpu to perform well, it is not that useful

**Comment by KiloNovemberDelta** [comment]
Yes indeed, I have checked, CPU is very high and GPU not used. I hope a fix will be delivered soon.

**Comment by svetoslavenchev** [comment]
Looks like the latest Intel Graphics Driver 30.0.101.1404, 2022-03-08 finally fixes the issue.
Tested on latest Windows 11, Ubuntu 20.04, Mesa 21.3.7 (kisak-mesa stable PPA).

**Comment by williamnavaraj** [comment]
Any fix for Nvidia.

**Comment by tank104** [comment]
I have intel Iris Xe, and upgrade to:
Intel Graphics Driver 30.0.101.1404 and stil had not luck.  Even tried the later version than that.
Still get blank screen

Tested on latest Windows 11, Ubuntu 20.04, Mesa 21.3.8

**Comment by giulioturrisi** [comment]
Same problem on my dell xps 13 9310 with intel iris xe and n. driver 30.0.101.1660

**Comment by giulioturrisi** [comment]
> Does anyone with Intel Iris happen to dual boot Linux is I see if it’s just a WSL issue or in both environments?

Here! I have not problem on the linux partition

**Comment by rmorassi** [comment]
Hello everyone,
Looks like the latest Intel Graphics Driver 30.0.101.1960 (10th May 2022) fixes the issue for the Intel(R) Iris(R) Xe. Rviz and Gazebo are working well and I don't have any black screens on WSL2.

**Comment by giulioturrisi** [comment]
The same for me, the problem still persist with the new driver (dell xps 9310)

**Comment by tank104** [comment]
Anyone had any luck getting this working in hardware mode?

**Comment by oliveira-jeliel** [comment]
Hey guys, I'm also with the same problem, glxgears works fine showing gears, but with driver Intel® Graphics Driver 31.0.101.3222/31.0.101.2111 on Windows 11  Intel Iris Xe,  Gazebo just shows black screen, managed to get it working only on software mode `export  LIBGL_ALWAYS_SOFTWARE=1`. Anyone had success?

**Comment by duyipai** [comment]
> Hey guys, I'm also with the same problem, glxgears works fine showing gears, but with driver Intel® Graphics Driver 31.0.101.3222/31.0.101.2111 on Windows 11 Intel Iris Xe, Gazebo just shows black screen, managed to get it working only on software mode `export LIBGL_ALWAYS_SOFTWARE=1`. Anyone had success?

Same here. No success with 31.0.101.3222

**Comment by biiigwang** [comment]
My friends, when I use coppeliaSim, the same problem occurs🤷: some of the openGL based screens are black. Hope to repair it as soon as possible!🙏

**Comment by giulioturrisi** [comment]
Just to keep the issue alive, the problem is still here, not matter which kernel or driver I'm using!

**Comment by gulerburak** [comment]
I have i7-12700H, Iris Xe, latest version of Win11 and Ubuntu 20.04, and the latest driver 31.0.101.4032. Rviz2 works but gazebo is still black screen.

![image](https://user-images.githubusercontent.com/63360489/213911738-f4fcba11-927b-4c39-a772-f540f07fecd1.png)

**Comment by rosiakpiotr** [comment]
Any chances to make it work?

**Comment by DDDeluge** [comment]
I found this problem which is related to Intel HD Graphics, when disabling Intel(R) UHD Graphics 770 in Device Manager, gazebo can work correctly.🤔

WSL2: Ubuntu22.04
CPU: 12th Gen Intel(R) Core(TM) i7-12700
GPU0: Intel(R) UHD Graphics 770
GPU1: NVIDIA GeForce RTX 3070Ti

**Comment by rosiakpiotr** [comment]
Is there anything happening towards solution? Feeling like this has been dead for too long now, being crucial feature for WSLg users that develop robotic systems using Gazebo.

**Comment by ZacZhangzhuo** [comment]
> Looks like the latest Intel Graphics Driver 30.0.101.1404, 2022-03-08 finally fixes the issue. Tested on latest Windows 11, Ubuntu 20.04, Mesa 21.3.7 (kisak-mesa stable PPA).

saved my life TWO times!!!! The other day windows updated the system, then I have to reinstall this driver...

**Comment by plantroon** [comment]
Laptop with 1235U Intel Xe, still no way to get accelerated graphics in WSL. llvmpipeis CPU rendering, not a solution at all. My 5 years old Galaxy phone with Dex can run llvmpipe and just like the laptop, it barely plays FullHD youtube in Firefox. With working graphics acceleration it should be an easy task.

**Comment by zhangbaozhe** [comment]
Ubuntu 20.04 & Windows 11 & Intel Ultra 9 185H with Arc graphics still not working. Gazebo even won't show up unless using `export LIBGL_ALWAYS_SOFTWARE=1`.

**Comment by zhangbaozhe** [comment]
> Ubuntu 20.04 & Windows 11 & Intel Ultra 9 185H with Arc graphics still not working. Gazebo even won't show up unless using `export LIBGL_ALWAYS_SOFTWARE=1`.

After I type `gazebo` in the terminal, it gives me a `D3D12: Removing Device.` warning.

**Comment by icyreq** [comment]
I try everything ,finally `export GALLIUM_DRIVER=llvmpipe`  work!!

**Comment by MaHao777** [comment]
> > Ubuntu 20.04 & Windows 11 & Intel Ultra 9 185H with Arc graphics still not working. Gazebo even won't show up unless using `export LIBGL_ALWAYS_SOFTWARE=1`.Ubuntu 20.04、Windows 11 以及带 Arc 显卡的 Intel Ultra 9 185H 依然无法工作。凉亭甚至不会显示，除非使用 `export LIBGL_ALWAYS_SOFTWARE=1` 。
> 
> After I type `gazebo` in the terminal, it gives me a `D3D12: Removing Device.` warning.我在终端输入“`gazebo`”后，它会显示 `D3D12：正在移除设备。` 警告。

I also encountered the same issue and successfully resolved it by setting the environment variable `export GALLIUM_DRIVER=llvmpipe`. This method bypasses GPU acceleration and uses CPU rendering, thus solving the problem of the 3D space not rendering. Although this is not the most ideal solution, as it sacrifices performance, it does indeed allow you to display Gazebo and Rviz normally in the absence of GPU acceleration. I hope this method is also helpful to you. Or maybe there is a better solution which can use GPU.
